### PR TITLE
Make tf.case work in eager mode with dict inputs

### PR DESCRIPTION
**Pull request by frreiss**
`tf.case` does not work in eager mode if the caller passes an unordered dictionary of predicate function pairs. In particular, the example code in the API documentation for `tf.case` doesn't run in eager mode:

```
In [1]: import tensorflow as tf 
   ...: tf.enable_eager_execution() 
   ...: (x,y,z) = (1,2,3) 
   ...: def f1(): return tf.constant(17) 
   ...: def f2(): return tf.constant(23) 
   ...: def f3(): return tf.constant(-1) 
   ...: tf.case({tf.less(x, y): f1, tf.greater(x, z): f2}, 
   ...:         default=f3, exclusive=True)                                     
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
[stack trace]
~/temp/testenv/lib/python3.6/site-packages/tensorflow/python/framework/ops.py in name(self)
    950   def name(self):
    951     raise AttributeError(
--> 952         "Tensor.name is meaningless when eager execution is enabled.")
    953 
    954   @property

AttributeError: Tensor.name is meaningless when eager execution is enabled.
```

The reason for this problem is that `tf.case` uses the names of the predicate ops to define the sort order of the predicates. This sort order is needed to make the op have deterministic behavior when `exclusive=False`.

This PR modifies `tf.case` so that, if the user passes an unordered dictionary for the `pred_fn_pairs`, and eager mode is enabled, and the `exclusive` argument is set to `True`; then `tf.case` will evaluate all the predicates in dictionary traversal order.

If `pred_fn_pairs` is an unordered dictionary, and eager mode is enabled, and the `exclusive` argument is `False`; the new code throws an exception. The traversal order of unordered dictionaries is not deterministic on versions of Python before 3.6.

I also updated the API docs and fixed a few typos in the docs.

### CODE REVIEW

File: tensorflow/python/ops/control_flow_ops_test.py [hunk 1]
```diff
@@ -936,7 +936,14 @@ def testCase_withoutDefault_oneCondition(self):
       with self.assertRaisesRegexp(errors.InvalidArgumentError, "Input error:"):
         sess.run(output, feed_dict={x: 4})
 
-
+  @test_util.run_in_graph_and_eager_modes
+  def testCase_dict_eagerMode(self):
```
Comment (akshaym) [inline tensorflow/python/ops/control_flow_ops_test.py]: Nit: remove _eagerMode from the name of this test.

File: tensorflow/python/ops/control_flow_ops.py [hunk 2]
```diff
@@ -3611,12 +3611,23 @@ def _case_verify_and_canonicalize_args(pred_fn_pairs, exclusive, name,
...
-    pred_fn_pairs = sorted(pred_fn_pairs.items(), key=lambda item: item[0].name)
-    if not exclusive:
-      logging.warn(
-          "%s: An unordered dictionary of predicate/fn pairs was "
-          "provided, but exclusive=False. The order of conditional "
-          "tests is deterministic but not guaranteed.", name)
+    if context.executing_eagerly():
+      # No name to sort on in eager mode. Use dictionary traversal order, 
+      # which is nondeterministic in versions of Python < 3.6
+      if not exclusive and not isinstance(pred_fn_pairs,
```
Comment (akshaym) [inline tensorflow/python/ops/control_flow_ops.py]: Minor nit: I don't think you need the "isinstance(pred_fn_pairs, collections.OrderedDict)".

**Review by akshaym — COMMENTED** [review]
Thanks for the contribution!

File: tensorflow/python/ops/control_flow_ops_test.py (hunk 1 above)
Comment (frreiss) [reply tensorflow/python/ops/control_flow_ops_test.py]: Removed _eagerMode from the name

File: tensorflow/python/ops/control_flow_ops.py (hunk 2 above)
Comment (frreiss) [reply tensorflow/python/ops/control_flow_ops.py]: Oops! I didn't notice the if statement 4 lines up :-\ Removed that predicate.

**Comment by frreiss** [comment]
Pushed some updates to address review comments.

**Review by akshaym — APPROVED** [review]

**Pull request merged**